### PR TITLE
.Net Processes - Map Step Feature

### PR DESCRIPTION
**Pull request by crickman**
### Motivation and Context

Fixes: https://github.com/microsoft/semantic-kernel/issues/9193

### Description

Map each value from a set to a map-operation and present the results as a set for potential reduction.  

**Includes:**
- `ProcessMapBuilder` (Core)
- `KernelProcessMap` / `KernelProcessMapState` (Abstractions)
- `LocalMap` (LocalRuntime)
- `MapActor` (DaprRuntime)
- Serialization
- Unit Tests (exhaustive)
- Integration Tests
- Sample

**Features:**
- Executes map operations in parallel
- Handles when output type has been transformed from input type
- Accepts either step or subprocess for map operation
- Proxies events for edges defined directly on the map-step
- Allows additional (non proxied) edges for the map-operation
- Participates in state serialization

**Follow-up:**
- Documentation - https://github.com/microsoft/semantic-kernel/issues/9616

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Experimental/Process.Core/ProcessMapBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            throw new KernelException("The target function must have a parameter name.");
+        }
+
+        // Build the edges first
+        var builtEdges = this.Edges.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.Select(e => e.Build()).ToList());
+        ProcessBuilder mapProcess = this.MapProcess;
+
+        KernelProcessMapState state = new(this.Name, this.Id);
+        return new KernelProcessMap(state, mapProcess.Build(), this.TargetFunction.ParameterName!, builtEdges);
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.Core/ProcessMapBuilder.cs:90]: thinking about state propagation at this stage there may need a state validation to either pass the state or pass null (like no state was given) before letting the build logic take care of it.

Also thinking about if it's needed to have a stepId here or just keep treating the map steps as an array

**Comment by esttenorio** [comment]
General thoughts about State + Map:

Since the map is a subprocess with N steps (to be defined at runtime), I'm thinking of 2 scenarios:

- Map steps are volatile and no state is saved, and if the root process pauses while running the map, the last event received by the map gets re-propagated and the map steps get re-spawned -> only works if map steps steps are not too complex
- Assuming the Map steps are complex enough (another subprocess as step + state) different strategies could be taken, since at this point the map steps act as workers -> trying to get multiple jobs (same jobs) done, and return a result:
    - all Map steps steps are singletons and share same resources, so on pause/resume only 1 state per map step step is saved -> using as example the foodprep -> if 5 fish orders get triggered the ingredient stock should be globally -5
         - also the singleton step may be an interesting feature to explore when reusing steps so they all use same resources just used multiple times by different processes, emitting result to different parent processes
    - on pause, the most completed process state gets saved, and on resume this one gets passed as initial state to all map steps:
       - with this consideration, when loading a processes with state at most for map steps we offer to pass along the same initial state to all map children

**Comment by esttenorio** [comment]
Additional considerations - not for this PR but for the feature in general:

- **timeout management**: if spawning N events, have timeout logic and if some map steps are not done, null is returned
- **limiting number of N elements**: depending on how complex the inner steps are (nested subprocesses)
- **volatile state**: since map steps are spawned on runtime have checkpoints before and after the map step

**Comment by crickman** [comment]
> Additional considerations - not for this PR but for the feature in general:
> 
> * **timeout management**: if spawning N events, have timeout logic and if some map steps are not done, null is returned
> * **limiting number of N elements**: depending on how complex the inner steps are (nested subprocesses)
> * **volatile state**: since map steps are spawned on runtime have checkpoints before and after the map step

I like these considerations.  I feel that _timeout_ and _depth_ considerations apply generally to the entire framework, not just the _map-step_.

I very much like the _volatile_ concept.  I have to admit, I'm not entirely clear on the pause or cancel mechanics in the runtime.  I don't see the cancel-token being propagated much, but I need to take a closer look

**Comment by crickman** [comment]
>   * all Map steps steps are singletons and share same resources, so on pause/resume only 1 state per map step step is saved -> using as example the foodprep -> if 5 fish orders get triggered the ingredient stock should be globally -5
>     
>     * also the singleton step may be an interesting feature to explore when reusing steps so they all use same resources just used multiple times by different processes, emitting result to different parent processes
>   * on pause, the most completed process state gets saved, and on resume this one gets passed as initial state to all map steps:

Ben and I did discuss this "functoid" constrain for the map operation and felt it might be too limiting.  At this point, the feature is functional without constraining the map operation state.  Also if we apply the "volatile" concept, I wonder if the need exists to constrain the map operation.

Another question that comes to mind is how to enforce such a constraint (singleton / functoid).

File: dotnet/src/Experimental/Process.Core/ProcessMapBuilder.cs (hunk 1 above)
Comment (crickman) [reply dotnet/src/Experimental/Process.Core/ProcessMapBuilder.cs]: I think your current approach on managinge state may address this.  Will merge today and verify.

File: dotnet/samples/Demos/ProcessWithDapr/Controllers/ProcessController.cs [hunk 2]
```diff
@@ -32,13 +32,195 @@ public ProcessController(Kernel kernel)
...
         var process = this.GetProcess();
-        var processContext = await process.StartAsync(new KernelProcessEvent() { Id = CommonEvents.StartProcess }, processId: processId);
+        var processContext = await process.Build().StartAsync(new KernelProcessEvent() { Id = CommonEvents.StartProcess, Data = 3 }, processId);
         var finalState = await processContext.GetStateAsync();
 
         return this.Ok(processId);
     }
 
-    private KernelProcess GetProcess()
+    /// <summary>
```
Comment (crickman) [inline dotnet/samples/Demos/ProcessWithDapr/Controllers/ProcessController.cs:41]: Will revert these changes for final.

File: dotnet/samples/Demos/ProcessWithDapr/Program.cs [hunk 3]
```diff
@@ -12,7 +12,8 @@
 });
 
 // Configure the Kernel with DI. This is required for dependency injection to work with processes.
-builder.Services.AddKernel();
+//builder.Services.AddKernel();
+builder.Services.AddSingleton(new Kernel());
```
Comment (crickman) [inline dotnet/samples/Demos/ProcessWithDapr/Program.cs:16]: Will revert for final.

File: dotnet/src/Experimental/Process.Core/ProcessStepBuilder.cs [hunk 4]
```diff
@@ -66,6 +66,8 @@ public ProcessStepEdgeBuilder OnFunctionError(string functionName)
 
     #endregion
 
+    internal const char EventIdSeparator = '.';
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.Core/ProcessStepBuilder.cs:69]: [nit] remove since it seems unused (?)

File: dotnet/src/Experimental/Process.LocalRuntime/LocalKernelProcessFactory.cs [hunk 5]
```diff
@@ -18,12 +18,9 @@ public static class LocalKernelProcessFactory
     /// <returns>An instance of <see cref="KernelProcess"/> that can be used to interrogate or stop the running process.</returns>
     public static async Task<LocalKernelProcessContext> StartAsync(this KernelProcess process, Kernel kernel, KernelProcessEvent initialEvent)
     {
-        Verify.NotNull(process);
-        Verify.NotNullOrWhiteSpace(process.State?.Name);
-        Verify.NotNull(kernel);
-        Verify.NotNull(initialEvent);
+        Verify.NotNull(initialEvent, nameof(initialEvent));
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalKernelProcessFactory.cs:21]: in general wondering if for initialEvent there should also be a check for the Id not not be empty

File: dotnet/src/Experimental/Process.UnitTests/KernelProcessSerializationTests.cs [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+using Microsoft.SemanticKernel.Process.Models;
+using Xunit;
+
+namespace Microsoft.SemanticKernel.Process.UnitTests;
+
+/// <summary>
+/// Unit testing of <see cref="Microsoft.SemanticKernel.Process.Models"/>
+/// and associated operations.
+/// </summary>
+public class KernelProcessSerializationTests
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.UnitTests/KernelProcessSerializationTests.cs:15]: nice! thank you for adding these!

File: dotnet/src/Experimental/Process.IntegrationTests.Shared/ProcessMapTests.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    /// </summary>
+    /// <returns>A <see cref="Task"/></returns>
+    [Fact]
+    public async Task TestMapAsync()
+    {
+        // Arrange
+        ProcessBuilder process = new(nameof(TestMapAsync));
+
+        ProcessStepBuilder computeStep = process.AddStepFromType<ComputeStep>();
+        ProcessMapBuilder mapStep = process.AddMapForTarget(new ProcessFunctionTargetBuilder(computeStep));
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.IntegrationTests.Shared/ProcessMapTests.cs:41]: [brainstorming hypothetical scenario]
wondering if these 2 should be a single function? since compute is not actually used by the root process directly having it instantiated on the level computeStep could in theory also be used as a regular step:


```mermaid
graph TD
root-->map-->compute1["compute"]
root-->compute2["compute"]
```

File: dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    /// Initializes a new instance of the <see cref="LocalMap"/> class.
+    /// </summary>
+    /// <param name="map">The <see cref="KernelProcessMap"/> instance.</param>
+    /// <param name="kernel">An instance of <see cref="Kernel"/></param>
+    internal LocalMap(KernelProcessMap map, Kernel kernel)
+        : base(map, kernel)
+    {
+        this._map = map;
+        this._logger = this._kernel.LoggerFactory?.CreateLogger(this._map.State.Name) ?? new NullLogger<LocalStep>();
+        this._mapEvents = [.. map.Edges.Keys.Select(key => key.Split('.').Last())];
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs:31]: was this supposed to use the `EventIdSeparator` I saw somewhere?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            {
+                ++index;
+
+                KernelProcess process = this._map.Operation.CloneProcess(this._logger);
+                MapOperationContext context = new(this._mapEvents, capturedEvents);
+#pragma warning disable CA2000 // Dispose objects before losing scope
+                LocalKernelProcessContext processContext = new(process, this._kernel, context.Filter);
+                Task processTask =
+                    processContext.StartWithEventAsync(
+                        new KernelProcessEvent
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs:55]: so all events emitted from a map are internal right? maybe if one was marked as public at this level a warning can be raised so people know

File: dotnet/src/Experimental/Process.IntegrationTests.Shared/ProcessMapTests.cs (hunk 7 above)
Comment (crickman) [reply dotnet/src/Experimental/Process.IntegrationTests.Shared/ProcessMapTests.cs:41]: Good question.  Ben and I were discussing this also.  There is a case where an edge may be defined outside of the map-operation: https://github.com/microsoft/semantic-kernel/blob/processes-map-step/dotnet/src/Experimental/Process.UnitTests/Runtime.Local/LocalMapTests.cs#L295

```mermaid
graph TD
root["root process"]-->map
subgraph map
compute1["compute"]
end
compute1["compute"]--1-->orthogonal
compute1["compute"]--N-->union
root-->compute2["compute"]
```

File: dotnet/src/Experimental/Process.UnitTests/Core/ProcessMapBuilderTests.cs [hunk 10]
```diff
@@ -0,0 +1,223 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (esttenorio) [inline dotnet/src/Experimental/Process.UnitTests/Core/ProcessMapBuilderTests.cs:1]: wondering how the error messages are dealt with the map step?
same as you get an [] of nullable results, when subscribing to onError you get an [] of errors?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs (hunk 9 above)
Comment (crickman) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs:55]: Correct, events specific to the map-operation (map-step edges) are internal as they target the process in which the map step is defined.  Any other proxied event will retain its original visibility.

File: dotnet/src/Experimental/Process.UnitTests/Core/ProcessMapBuilderTests.cs (hunk 10 above)
Comment (crickman) [reply dotnet/src/Experimental/Process.UnitTests/Core/ProcessMapBuilderTests.cs:1]: yes, I believe so.  if out of 5 steps 3 result in some type of error and 2 don't, three invocations of the error handler seems natural.  i'm hestant to intercept and interpret external failure.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalKernelProcessFactory.cs (hunk 5 above)
Comment (crickman) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalKernelProcessFactory.cs:21]: This is a good point.  I've updated `KernelProcessEvent.Id` so its no longer nullable, but it might be whitespace or empty string.  I suppose my thought is that this condition is relevant for any event that is emitted, so perhaps the check might be more useful as part of the processing loop.  I've updated `LocalProcess.RunOnceAsync` to check for this (as opposed to the factory) and also `LocalStep.EmitEventAsync` and `KernelProcessStepContext.EmitEventAsync`.

File: dotnet/src/Experimental/Process.Abstractions/KernelProcessMap.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+/// </summary>
+public sealed record KernelProcessMap : KernelProcessStepInfo
+{
+    /// <summary>
+    /// Event Id used internally to initiate the map operation.
+    /// </summary>
+    public const string MapEventId = "StartMap";
+
+    /// <summary>
+    /// The map operation.
```
[the remark this answers is not in the document]
Comment (crickman) [reply dotnet/src/Experimental/Process.Abstractions/KernelProcessMap.cs:17]: Now that we have our constants file, I've moved this definition (and updated comment).

File: dotnet/src/Experimental/Process.Core/ProcessStepBuilder.cs (hunk 4 above)
Comment (crickman) [reply dotnet/src/Experimental/Process.Core/ProcessStepBuilder.cs:69]: Using now, as you identified.

**Review by esttenorio — APPROVED** [review]

File: dotnet/src/Experimental/Process.IntegrationTests.Shared/ProcessMapTests.cs (hunk 7 above)
Comment (crickman) [reply dotnet/src/Experimental/Process.IntegrationTests.Shared/ProcessMapTests.cs:41]: Worked with Ben to evolve what's going on here

File: dotnet/samples/GettingStartedWithProcesses/Step05/Step05_MapReduce.cs [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+        this._sourceContent = content.ToString().ToUpperInvariant();
+    }
+
+    [Fact]
+    public async Task RunMapReduceAsync()
+    {
+        KernelProcess process = SetupMapReduceProcess(nameof(RunMapReduceAsync), "Start");
+        Kernel kernel = new();
+        using LocalKernelProcessContext localProcess =
```
Comment (TaoChenOSU) [inline dotnet/samples/GettingStartedWithProcesses/Step05/Step05_MapReduce.cs:42]: For my own education: why do we need this context here. I don't see it getting used.

File: dotnet/src/Experimental/Process.Abstractions/Models/KernelProcessMapStateMetadata.cs [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+/// <summary>
+/// Process state used for State Persistence serialization
+/// </summary>
+public sealed record class KernelProcessMapStateMetadata : KernelProcessStepStateMetadata
+{
+    /// <summary>
+    /// Process State of Steps if provided
+    /// </summary>
+    [DataMember]
+    [JsonPropertyName("mapState")]
```
Comment (TaoChenOSU) [inline dotnet/src/Experimental/Process.Abstractions/Models/KernelProcessMapStateMetadata.cs:16]: Should this be `mapState` or `operationState`?

File: dotnet/samples/GettingStartedWithProcesses/Step05/Step05_MapReduce.cs (hunk 12 above)
Comment (crickman) [reply dotnet/samples/GettingStartedWithProcesses/Step05/Step05_MapReduce.cs:43]: This is a feature of how the framework is designed/implemented.  One _could_ use this context to cancel or interact with the process.   We're not doing anything here...but since its disposable it needs to be handled.

File: dotnet/src/Experimental/Process.Abstractions/Models/KernelProcessMapStateMetadata.cs (hunk 13 above)
Comment (crickman) [reply dotnet/src/Experimental/Process.Abstractions/Models/KernelProcessMapStateMetadata.cs]: Good eye, yes.  There was some churn around this on merging updates.

**Review by TaoChenOSU — APPROVED** [review]

File: dotnet/src/Experimental/Process.Runtime.Dapr/Interfaces/IMap.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    /// </summary>
+    /// <returns>A <see cref="Task"/></returns>
+    /// <exception cref="KernelException"></exception>
+    Task InitializeMapAsync(DaprMapInfo mapInfo, string? parentProcessId);
+
+    ///// <summary>
+    ///// Builds the current state of the step into a <see cref="DaprMapInfo"/>.
+    ///// </summary>
+    ///// <returns>An instance of <see cref="DaprMapInfo"/></returns>
+    //Task<DaprMapInfo> ToDaprMapInfoAsync();
```
[the remark this answers is not in the document]
Comment (crickman) [reply dotnet/src/Experimental/Process.Runtime.Dapr/Interfaces/IMap.cs]: Already removed (was doing a final look over when you pinged on review)

File: dotnet/src/Experimental/Process.Runtime.Dapr/Interfaces/IMap.cs [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+/// An interface that represents a step in a process.
+/// </summary>
+public interface IMap : IActor
+{
+    /// <summary>
+    /// Initializes the step with the provided step information.
+    /// </summary>
+    /// <returns>A <see cref="Task"/></returns>
+    /// <exception cref="KernelException"></exception>
+    Task InitializeMapAsync(DaprMapInfo mapInfo, string? parentProcessId);
```
[the remark this answers is not in the document]
Comment (crickman) [reply dotnet/src/Experimental/Process.Runtime.Dapr/Interfaces/IMap.cs:17]: It is a dapr actor interface.  Following the pattern Ben has established.  Happy to walk through my understanding of dapr actors if helpful.

**Review by TaoChenOSU — APPROVED** [review]

File: dotnet/src/Experimental/Process.LocalRuntime/LocalKernelProcessContext.cs [hunk 16]
```diff
@@ -14,17 +13,22 @@ public sealed class LocalKernelProcessContext : KernelProcessContext, IDisposabl
     private readonly LocalProcess _localProcess;
     private readonly Kernel _kernel;
 
-    internal LocalKernelProcessContext(KernelProcess process, Kernel kernel)
+    internal LocalKernelProcessContext(KernelProcess process, Kernel kernel, ProcessEventProxy? filter = null)
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalKernelProcessContext.cs:16]: Would `eventProxy` be a better name for the parameter? Given how it is used if a developer is using named parameters it might be clearer.

Comment (crickman) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalKernelProcessContext.cs]: Yes, agree.  Updated

File: dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+                    }
+
+                    results.SetValue(result, index);
+                }
+            }
+
+            // Emit map results
+            foreach (string eventName in capturedEvents.Keys)
+            {
+                Array eventResult = resultMap[eventName];
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs:93]: `eventResults` would be a better name

Comment (crickman) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs:93]: I hear where you are coming from.  The cardinality aspects at play here a bit complex.  Each map-step "result" is always an enumeration and a map operation may emit multiple events or "results"...which would be multiple enumerations.

This is a single enumeration or "result" that is associated with a single event.  The for loop is emitting multiple "results" (as commented).

**Review by TaoChenOSU — APPROVED** [review]

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 18]
```diff
@@ -76,6 +75,11 @@ public LocalStep(KernelProcessStepInfo stepInfo, Kernel kernel, string? parentPr
     /// </summary>
     internal string Id => this._stepInfo.State.Id!;
 
+    /// <summary>
+    /// An event filter that can be used to intercept events emitted by the step.
+    /// </summary>
+    internal ProcessEventProxy? EventProxy { get; init; }
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:81]: Following up on my early comment, it looks odd that the description uses filter but the name uses proxy.

If I understand the usage, this is a delegate that gets invoked before emitting an event, which if false is returned will prevent the event being emitted.

Consider rethinking the naming here, maybe something like:

```csharp
ShouldEmitEventHandler? ShouldEmitEvent { ... }
ShouldEmitEventDelegate? ShouldEmitEvent { ... }

Comment (crickman) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:81]: Updated comment.  Creating issue to continue evaluation of the internal event-proxy feature.  We're considering how this might apply outside of map-step, which will provide additional data-points to drive evolution.

**Review by TaoChenOSU — APPROVED** [review]

**Review by markwallace-microsoft — APPROVED** [review]
